contracts/dev-utils/contracts/src/LibAssetData.sol

```diff
@@ -0,0 +1,518 @@
+            for (uint256 i = 0; i != length; i++) {
+                // Query balance of individual assetData
+                uint256 totalBalance = getBalance(ownerAddress, nestedAssetData[i]);
+
+                // Scale total balance down by corresponding value in assetData
+                uint256 scaledBalance = totalBalance / assetAmounts[i];
+                if (scaledBalance < balance || balance == 0) {
+                    balance = scaledBalance;
+                }
+            }
+        } 
+
+        // Balance will be 0 if assetProxyId is unknown
+        return balance;
+    }
+
+    /// @dev Calls getBalance() for each element of assetData.
+    /// @param ownerAddress Owner of the assets specified by assetData.
+    /// @param assetData Array of asset details, each encoded per the AssetProxy contract specification.
+    /// @return Array of asset balances from getBalance(), with each element
+    /// corresponding to the same-indexed element in the assetData input.
+    function getBatchBalances(address ownerAddress, bytes[] memory assetData)
+        public
+        view
+        returns (uint256[] memory balances)
+    {
+        uint256 length = assetData.length;
+        balances = new uint256[](length);
+        for (uint256 i = 0; i != length; i++) {
+            balances[i] = getBalance(ownerAddress, assetData[i]);
+        }
+        return balances;
+    }
+
+    /// @dev Returns the number of asset(s) (described by assetData) that
+    /// the corresponding AssetProxy contract is authorized to spend.  When the asset data contains
+    /// multiple assets (eg for Multi-Asset), the return value indicates
+    /// how many complete "baskets" of those assets may be spent by all of the corresponding
+    /// AssetProxy contracts.
+    /// @param ownerAddress Owner of the assets specified by assetData.
+    /// @param assetData Details of asset, encoded per the AssetProxy contract specification.
+    /// @return Number of assets (or asset baskets) that the corresponding AssetProxy is authorized to spend.
+    function getAssetProxyAllowance(address ownerAddress, bytes memory assetData)
+        public
+        view
+        returns (uint256 allowance)
+    {
+        // Get id of AssetProxy contract
+        bytes4 assetProxyId = assetData.readBytes4(0);
+
+        if (assetProxyId == MULTI_ASSET_PROXY_ID) {
+            // Get array of values and array of assetDatas
+            (, uint256[] memory amounts, bytes[] memory nestedAssetData) = decodeMultiAssetData(assetData);
+
+            uint256 length = nestedAssetData.length;
+            for (uint256 i = 0; i != length; i++) {
+                // Query allowance of individual assetData
+                uint256 totalAllowance = getAssetProxyAllowance(ownerAddress, nestedAssetData[i]);
+
+                // Scale total allowance down by corresponding value in assetData
+                uint256 scaledAllowance = totalAllowance / amounts[i];
```

Comment:
I wonder if we should special case this for `isApprovedForAll`. Suppose the nested assets are all ERC1155 with `isApprovedForAll=true`. A client may expect the returned allowance to be `_MAX_UINT256`.

Reply:
It's an interesting idea, but it does introduce another edge case where `order.takerAssetAmount * allowance` could overflow. Might still make this change at some point, but I think I'll punt on it for now.